### PR TITLE
Command Palette: Track select event when current site is assumed

### PR DESCRIPTION
**Pull request by mmtr**
## Proposed Changes

Fixes a regression that was preventing the `calypso_hosting_command_palette_command_select` event from being tracked on commands when the current site is assumed.

## Testing Instructions

- Use the Calypso live link below
- Go to a site-specific page (e.g. `/home/:site`)
- Open the command palette
- Select a command with multi-site support (e.g. "Open hosting configuration")
- Make sure the `calypso_hosting_command_palette_command_select` event is tracked
- Make sure the command redirects to the appropriate path for the current site without showing the site selector
- Make sure the `calypso_hosting_command_palette_site_select` event is NOT tracked
- Go to a siteless page (e.g. `/read`)
- Open the command palette
- Select a command with multi-site support (e.g. "Open hosting configuration")
- Make sure the `calypso_hosting_command_palette_command_select` event is tracked
- Make sure the site selector shows up
- Select a site
- Make sure the `calypso_hosting_command_palette_site_select` event is tracked
- Make sure the commands redirects to the appropriate path for the selected site

### CODE REVIEW

**Comment by rcrdortiz** [comment]
This PR works as advertised. But I'm thinking that it might make sense to change the events a bit. Right now, on a siteless command, these are the triggered events after completing the normal flow:
- calypso_hosting_command_palette_open
- calypso_hosting_command_palette_command_select
- calypso_hosting_command_palette_site_select
- calypso_hosting_command_palette_close

If we open the command palette, select a command and bail, the flow is:
- calypso_hosting_command_palette_open
- calypso_hosting_command_palette_command_select
- calypso_hosting_command_palette_close

When we have a selected site the flow is:
- calypso_hosting_command_palette_open
- calypso_hosting_command_palette_command_select
- calypso_hosting_command_palette_close

If we bail on a multi-site or complete a single-site command we have the same flow, this might lead to confusion if the person that is analyzing the events isn't aware of this. It might be better to have two different events, one of them could keep the existing `calypso_hosting_command_palette_command_select` and we could have a new one called (feel free to come up with a better name) `calypso_hosting_command_palette_command_execute`. What do you think?

**Review by rcrdortiz — APPROVED** [review]
The code looks good and tests well.

**Comment by mmtr** [comment]
> If we bail on a multi-site or complete a single-site command we have the same flow, this might lead to confusion if the person that is analyzing the events isn't aware of this.

That's a good point. Addressed in https://github.com/Automattic/wp-calypso/pull/88285/commits/1d417e9cf3a62c32caaea36a6767577bb44dc87d, although I don't particularly like the approach (we're duplicating stuff, I couldn't find any elegant way of avoiding it).

File: packages/command-palette/src/use-command-palette.tsx [hunk 1]
```diff
@@ -293,7 +318,18 @@ export const useCommandPalette = ( {
...
 		finalSortedCommands = finalSortedCommands.map( ( command: Command ) => {
 			const callback = ( params: CommandCallBackParams ) => {
-				const targetFunction = command?.siteFunctions?.onClick || command.callback;
+				let targetFunction;
+				const isMultiSiteCommand = command?.siteFunctions !== undefined;
+				if ( isMultiSiteCommand ) {
+					targetFunction = command?.siteFunctions?.onClick;
+					// We need to track the selected command event here because `command.siteFunctions.onClick`
+					// does not track anything (and it's not meant to).
+					trackSelectedCommand( command );
```
Comment (rcrdortiz) [inline packages/command-palette/src/use-command-palette.tsx:323]: This here needs to track the completed command. All this logical branch only gets executed when we have a selected current site. In that context what we're doing is mapping the multiSiteCommand to a singleSiteCommand and the only possible outcome is `trackCompletedCommand`.

Here we create an adapter function that adapts a multisite command to behave like a single-site command.

File: packages/command-palette/src/use-command-palette.tsx [hunk 2]
```diff
@@ -293,7 +318,18 @@ export const useCommandPalette = ( {
...
+				} else {
+					// We don't need to track the selected command event here because `command.callback`
+					// already does it (see how `finalSortedCommands` is created a few lines above).
+					targetFunction = command.callback;
```
Comment (rcrdortiz) [inline packages/command-palette/src/use-command-palette.tsx]: If the code reaches this point we should change the tracking from `isSelected` to `isCompleted`.  At line 301 we're always assigning  `isSelected` to the main callback.

File: packages/command-palette/src/use-command-palette.tsx (hunk 1 above)
Comment (mmtr) [reply packages/command-palette/src/use-command-palette.tsx:323]: I think I solved this in https://github.com/Automattic/wp-calypso/pull/88285/commits/c91689a61ceb9638e378425472f8937737f26e8f

File: packages/command-palette/src/use-command-palette.tsx (hunk 2 above)
Comment (mmtr) [reply packages/command-palette/src/use-command-palette.tsx]: I think I solved this in https://github.com/Automattic/wp-calypso/pull/88285/commits/c91689a61ceb9638e378425472f8937737f26e8f

File: packages/command-palette/src/use-command-palette.tsx [hunk 3]
```diff
@@ -188,6 +188,27 @@ export const useCommandPalette = ( {
...
+	const trackCompletedCommand = ( command: Command ) => {
+		recordTracksEvent( 'calypso_hosting_command_palette_command_complete', {
+			command: command.name,
+			list_count: commands.length,
+			list_visible_count: listVisibleCount,
+			current_route: currentRoute,
+			search_text: search,
+		} );
+	};
+
```
Comment (rcrdortiz) [inline packages/command-palette/src/use-command-palette.tsx:211]: [Nitpick]
We could improve this a bit by composing a higher-order function and having a base.
Something like this:
```
const trackCommand = (
		command: Command,
		eventName = 'calypso_hosting_command_palette_command_complete',
		extra = {}
	) => {
		recordTracksEvent( eventName, {
			command: command.name,
			list_count: commands.length,
			list_visible_count: listVisibleCount,
			current_route: currentRoute,
			search_text: search,
			...extra,
		} );
	};

	const trackSelectedCommand = ( command: Command ) => {
		trackCommand( command, 'calypso_hosting_command_palette_command_select', {
			has_nested_commands: !! command.siteFunctions,
		} );
	};
```
Or any other variation, like not creating `trackSelectedCommand` and explicitly calling the main function.

**Comment by rcrdortiz** [comment]
I'm getting some strange behavior. These are the steps I'm following:

1. Go to /sites
2. Open the command palette
3. click on hosting config
4. select a site

And these are the events I'm seeing:

1. calypso_hosting_command_palette_open
2. calypso_hosting_command_palette_command_select
3. calypso_hosting_command_palette_site_select
4. calypso_hosting_command_palette_navigate
5. calypso_hosting_command_palette_close
6. calypso_hosting_command_palette_command_complete

On single site commands the complete event happens before the close event. I don't think this should be a blocker but it's something strange that I noticed. Feel free to merge the PR since everything else is working as expected.

**Pull request merged**